lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -153,85 +184,35 @@ public void OnNext(IActiveContext activeContext)
         {
             Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", activeContext.Id));
 
-            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            if (!_reachedUpdateTaskActiveContext)
             {
+                Logger.Log(Level.Info, "Submitting master task");
                 _reachedUpdateTaskActiveContext = true;
-                RequestMapEvaluators(_dataSet.Count);
-
-                var partialTaskConf =
-                    TangFactory.GetTang()
-                        .NewConfigurationBuilder(new[]
-                        {
-                            TaskConfiguration.ConfigurationModule
-                                .Set(TaskConfiguration.Identifier,
-                                    IMRUConstants.UpdateTaskName)
-                                .Set(TaskConfiguration.Task,
-                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
-                                .Build(),
-                            _configurationManager.UpdateFunctionConfiguration,
-                            _configurationManager.ResultHandlerConfiguration
-                        })
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
-
-                try
-                {
-                    TangFactory.GetTang()
-                        .NewInjector(partialTaskConf, _configurationManager.UpdateFunctionCodecsConfiguration)
-                        .GetInstance<IIMRUResultHandler<TResult>>();
-                }
-                catch (InjectionException)
-                {
-                    partialTaskConf = TangFactory.GetTang().NewConfigurationBuilder(partialTaskConf)
-                        .BindImplementation(GenericType<IIMRUResultHandler<TResult>>.Class,
-                            GenericType<DefaultResultHandler<TResult>>.Class)
-                        .Build();
-                    Logger.Log(Level.Warning,
-                        "User has not given any way to handle IMRU result, defaulting to ignoring it");
-                }
-
+                var groupCommConfig = GetGroupCommConfiguration();
                 _commGroup.AddTask(IMRUConstants.UpdateTaskName);
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
+                var finalTaskConfig = Configurations.Merge(groupCommConfig, UpdateTaskConfiguration());
+                _groupCommTaskStarter.QueueTask(finalTaskConfig, activeContext);
+                RequestMapEvaluators(_dataSet.Count);
             }
             else
             {
+                Logger.Log(Level.Info, "Submitting map task");
+                _serviceAndContextConfigurationProvider.ReachedActiveContext(activeContext.EvaluatorId);
+                var groupCommConfig = GetGroupCommConfiguration();
                 string taskId;
-
                 if (!_taskIdStack.TryPop(out taskId))
```

Comment:
No there is no relation. And even in current code I do not think I use them together anywhere. I do not get the problem in using Stack. For fault tolerant case once we reach task creation stage you can just create the TaskId stack there and use it. You can use some other logic also like maintain integer for count etc. but I do not see any issue in using the either.